source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -1167,6 +1167,80 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         addUsersToWorkList(newStore);
     }
 
+    void processNonUniformResourceIndex(IRInst* nonUniformResourceIndexInst)
+    {
+        // implement the translation to spirv by waking up the use-def chain
+        // when we reach the leaf node (texture
+        // trace the use-def chain from nonuniformresource index def all the way to the
+        // textures / resources
+        // Decorate the following:
+        //   go through GEP and see if it calls an intrinsic function, decorate the address itself (GetElementPtr)
+        //   go through GEP to identify the pointer access and the Loads that it accesses (GetElementPtr -> Load)
+        //   go throuhg ExtractElement instruction for array access (GetElement)
+        //   go through IntCasts to deal with u32 -> i32 / vice-versa (IntCast)
+        // add the decoration to the first Use that is either an extract element  / a load of
+        // an extract element address instruction
+        // if the first use is an extract element ptr, find all the loads that uses the element,
+        // and the decorate them
+        //
+        // if the first use is a getelementptr, we may want to decorate the get element ptr itself,
+        // because user might
+        // call an image atomic operation, that operates on the ptr instead of the texture / resource
+        // todo rename the decoration itself nonuniformresourceindex -> nonuniformresource
+        List<IRInst*> resWorkList;
+        std::function<void(IRInst*)> visit;
+        visit = [&](IRInst* inst)
+        {
+            traverseUses(inst, [&](IRUse* use)
+            {
+                auto user = use->getUser();
+                IRBuilder builder(user);
+                builder.setInsertBefore(user);
+                IRInst* newUser = nullptr;
+                switch (user->getOp())
+                {
+                case kIROp_IntCast:
+                    printf("INTCAST!! %d\n", user->_debugUID);
+                    newUser = builder.emitCast(user->getFullType(), inst->getOperand(0));
+                    break;
+                case kIROp_GetElementPtr:
+                    // Replace gep(pArray, nonUniformRes(x)), into nonUniformRes(gep(pArray, x))
+                    printf("GETELEMENTPTR!! %d\n", user->_debugUID);
```

Comment:
You need to make sure the nonUniformResourxeIndex is on the index instead of the array. If it is on the array we don't want to push any further up.

gep(nonUniform(array), index) => no transform.

Reply:
i'm using a simple `if (user->getOperand(0)->getOp() != kIROp_NonUniformResourceIndex)` to check this. Is there a better way?